File: skills/architecture/scripts/utils/walk.mjs
import fs from 'node:fs';
import path from 'node:path';

/**
 * Recursively walk a directory, returning paths of TypeScript/JavaScript files.
 * @param {string} dir The directory to walk.
 * @param {string[]} exclude File/directory name substrings to exclude.
 * @param {Set<string>} [visited] Real paths already walked (cycle guard).
 * @returns {string[]} Resolved file paths.
 */
export function walkDir(dir, exclude = [], visited = new Set()) {
  let files = [];
  try {
    // Guard against symlink cycles (common in pnpm/yarn workspaces): track the
    // canonical path so we never recurse into the same real directory twice.
    let real;
    try {
      real = fs.realpathSync(dir);
    } catch {
      real = dir;
    }
    if (visited.has(real)) return files;
    visited.add(real);

    const list = fs.readdirSync(dir);
    for (const file of list) {
      if (exclude.some((ex) => file.includes(ex))) continue;
      const fullPath = path.join(dir, file);
      let stat;
      try {
        stat = fs.statSync(fullPath);
      } catch {
        // Skip files/dirs we can't read (permissions, symlinks, etc.)
        continue;
      }
      if (stat.isDirectory()) {
        files = files.concat(walkDir(fullPath, exclude, visited));
      } else if (
        fullPath.endsWith('.ts') ||
        fullPath.endsWith('.tsx') ||
        fullPath.endsWith('.js') ||
        fullPath.endsWith('.mjs') ||
        fullPath.endsWith('.py') ||
        fullPath.endsWith('.go')
      ) {
        files.push(fullPath);
      }
    }
  } catch {
    // Skip directories we can't read/traverse
  }
  return files;
}

File: skills/architecture/scripts/utils/walk.py
import os
from typing import List, Set, Optional


def walk_dir(
    root_dir: str,
    exclude: Optional[List[str]] = None,
    visited: Optional[Set[str]] = None,
) -> List[str]:
    """
    Recursively walk a directory, returning paths of TypeScript/JavaScript, Python, and Go files.
    """
    if exclude is None:
        exclude = []
    if visited is None:
        visited = set()

    files = []
    try:
        # Guard against symlink cycles
        real_path = os.path.realpath(root_dir)
        if real_path in visited:
            return files
        visited.add(real_path)

        if not os.path.exists(root_dir):
            return files

        for entry in os.scandir(root_dir):
            if any(pat in entry.name for pat in exclude):
                continue

            if entry.is_dir(follow_symlinks=True):
                files.extend(walk_dir(entry.path, exclude, visited))
            elif entry.is_file():
                if any(
                    entry.name.endswith(ext)
                    for ext in [".ts", ".tsx", ".js", ".mjs", ".py", ".go"]
                ):
                    files.append(os.path.abspath(entry.path))
    except (PermissionError, OSError):
        # Skip directories we can't read/traverse
        pass

    return files


if __name__ == "__main__":
    import sys
    import json

    test_dir = sys.argv[1] if len(sys.argv) > 1 else "."
    print(json.dumps(walk_dir(test_dir, exclude=["node_modules", ".git"]), indent=2))
